File: code/examples/node/12_langchain_agents.js
/**
 * 12_langchain_agents.js
 * =======================
 * LangChain Agents —— 让 AI 自主决定用什么工具完成任务
 *
 * 什么是 Agent？
 *   普通 AI：你问一个问题，它回答一个问题（被动响应）
 *   Agent：你给它一个目标，它自己决定"先搜资料、再计算、再总结"（主动规划）
 *
 *   类比：普通 AI 是"执行员"（给什么做什么），
 *         Agent 是"项目经理"（给目标，自己排工序）
 *
 * 涵盖内容：
 *   1. DynamicTool —— 简单工具定义（单字符串输入）
 *   2. DynamicStructuredTool —— 带 Zod Schema 的结构化工具（多参数）
 *   3. createReactAgent —— ReAct 模式 Agent 构建
 *   4. AgentExecutor —— 运行 Agent 并控制循环次数
 *   5. BufferMemory —— 对话记忆（跨轮次记住上下文）
 *   6. 流式 Agent 输出（streamEvents）
 *
 * 运行前：
 *   npm install @langchain/anthropic @langchain/core langchain zod
 *   export ANTHROPIC_API_KEY=sk-ant-...
 *   node 12_langchain_agents.js
 */

import { ChatAnthropic } from '@langchain/anthropic';
import { DynamicTool, DynamicStructuredTool } from '@langchain/core/tools';
import { createReactAgent } from 'langchain/agents';
import { AgentExecutor } from 'langchain/agents';
import { BufferMemory } from 'langchain/memory';
import { ChatPromptTemplate, MessagesPlaceholder } from '@langchain/core/prompts';
import { z } from 'zod';

// ─────────────────────────────────────────────
// 初始化模型
// ─────────────────────────────────────────────

const model = new ChatAnthropic({
  model: 'claude-opus-4-5',
  maxTokens: 2048,
  apiKey: process.env.ANTHROPIC_API_KEY,
});

// ─────────────────────────────────────────────
// 1. DynamicTool —— 最简单的工具定义方式
// ─────────────────────────────────────────────

/**
 * DynamicTool 的三要素：
 *   - name：工具名（AI 通过名字选择工具）
 *   - description：工具说明（AI 根据描述判断何时用）
 *   - func：实际执行的函数（接受字符串，返回字符串）
 *
 * 类比：给新员工配备工具箱，每个工具都贴着标签说明用途
 *
 * DynamicTool 的限制：输入只能是一个字符串（简单场景够用）
 * 需要多个参数时，用下面的 DynamicStructuredTool
 */

// 工具1：获取当前时间
const getCurrentTimeTool = new DynamicTool({
  name: 'get_current_time',
  description: '获取当前的日期和时间。当用户询问现在几点、今天是几号时使用。',
  func: async () => {
    return `当前时间：${new Date().toLocaleString('zh-CN', { timeZone: 'Asia/Shanghai' })}`;
  },
});

// 工具2：天气查询（模拟）
const getWeatherTool = new DynamicTool({
  name: 'get_weather',
  description: '查询某个城市的天气。输入城市名称，返回天气信息。',
  func: async (city) => {
    const weatherData = {
      北京: '晴，23°C，微风',
      上海: '多云，26°C，东南风3级',
      广州: '阵雨，28°C，南风4级',
      深圳: '阴，27°C，偏南风',
    };
    return `${city}天气：${weatherData[city] || `${city}：晴，22°C，微风（模拟数据）`}`;
  },
});

// ─────────────────────────────────────────────
// 2. DynamicStructuredTool —— 带 Zod Schema 的结构化工具
// ─────────────────────────────────────────────

/**
 * DynamicStructuredTool 的优势：
 *   支持多个类型化参数，AI 会自动解析参数格式。
 *   用 Zod 定义 Schema，内置类型验证。
 *
 * 与 DynamicTool 的对比：
 *   DynamicTool：   func(inputString) → 只有一个字符串参数
 *   DynamicStructuredTool：func({ param1, param2, ... }) → 结构化多参数
 *
 * 类比：
 *   DynamicTool 是"写个便签"（一句话说清楚）
 *   DynamicStructuredTool 是"填写结构化表单"（每个字段有类型和描述）
 */
const calculatorTool = new DynamicStructuredTool({
  name: 'calculator',
  description: '进行数学计算。支持加减乘除和括号，如 "2 + 3 * 4"。',
  schema: z.object({
    expression: z.string().describe('数学表达式，只含数字和 +、-、*、/、()'),
  }),
  func: async ({ expression }) => {
    try {
      const sanitized = expression.replace(/[^0-9+\-*/().\s]/g, '');
      // 注意：eval 有安全风险，生产环境用 math.js 或 mathjs
      const result = Function(`"use strict"; return (${sanitized})`)();
      return `${expression} = ${result}`;
    } catch {
      return `计算失败：表达式无效（${expression}）`;
    }
  },
});

const createReminderTool = new DynamicStructuredTool({
  name: 'create_reminder',
  description: '创建一个提醒事项，需要提供内容和时间。',
  schema: z.object({
    content: z.string().describe('提醒的具体内容'),
    time: z.string().describe('提醒时间，如 "明天上午9点" 或 "2025-01-01 09:00"'),
    priority: z.enum(['low', 'medium', 'high']).optional().describe('优先级（可选）'),
  }),
  func: async ({ content, time, priority = 'medium' }) => {
    const reminder = {
      id: Math.random().toString(36).slice(2, 8),
      content,
      time,
      priority,
      createdAt: new Date().toISOString(),
    };
    return `提醒已创建：${JSON.stringify(reminder)}`;
  },
});

const tools = [getCurrentTimeTool, getWeatherTool, calculatorTool, createReminderTool];

// ─────────────────────────────────────────────
// 3. createReactAgent + AgentExecutor
// ─────────────────────────────────────────────

/**
 * ReAct（Reasoning + Acting）模式
 *
 * Agent 的内部思考循环：
 *   1. Thought（推理）：我要完成什么，需要用哪个工具？
 *   2. Action（行动）：调用工具
 *   3. Observation（观察）：工具返回了什么？
 *   4. 重复，直到有足够信息给出最终答案
 *
 * 类比：侦探破案
 *   推理 → 调查 → 观察证据 → 再推理 → ... → 最终结论
 *
 * AgentExecutor 的作用：
 *   管理 ReAct 循环，限制最大迭代次数，避免死循环
 */

const agentSystemPrompt = `你是一个能干的 AI 助手，可以使用工具来完成任务。
请用中文回答，并在必要时使用工具获取准确信息。
当你有足够信息时，直接给出最终答案，不要过度调用工具。`;

async function demo1_basicAgent() {
  console.log('\n=== 1. 基础 ReAct Agent ===');

  const agent = await createReactAgent({
    llm: model,
    tools,
    messageModifier: agentSystemPrompt,
  });

  const executor = new AgentExecutor({
    agent,
    tools,
    verbose: true,     // 打印每步推理过程（调试用）
    maxIterations: 5,  // 最多循环 5 次
  });

  // 这个任务需要用到多个工具（时间 + 计算）
  const result = await executor.invoke({
    input: '现在几点了？顺便帮我算一下 365 * 24 * 60，这是一年有多少分钟。',
  });

  console.log('\n最终答案:', result.output);
}

// ─────────────────────────────────────────────
// 4. BufferMemory —— 对话记忆
// ─────────────────────────────────────────────

/**
 * BufferMemory
 *
 * 什么是"记忆"？
 *   默认 Agent 每次调用是独立的，不记得上一次说了什么。
 *   BufferMemory 把历史消息保存在内存里，每次调用时一起传给 AI。
 *
 *   类比：普通 AI 是"金鱼记忆"（每次聊天从头开始），
 *         加了 BufferMemory 的 AI 是"有笔记本的学生"（记得之前说过什么）
 *
 * 关键配置：
 *   - memoryKey     : Prompt 中对应的变量名（{chat_history}）
 *   - returnMessages: true → 返回 Message 对象（配合 MessagesPlaceholder 使用）
 *   - inputKey/outputKey: 告诉 Memory 哪些字段是输入/输出
 */
async function demo2_agentWithMemory() {
  console.log('\n=== 2. 带 BufferMemory 对话记忆的 Agent ===');

  const memory = new BufferMemory({
    memoryKey: 'chat_history',  // 对应 Prompt 中的 {chat_history}
    returnMessages: true,
    inputKey: 'input',
    outputKey: 'output',
  });

  // 带记忆的 Prompt 必须包含 MessagesPlaceholder 注入历史消息
  const promptWithMemory = ChatPromptTemplate.fromMessages([
    ['system', agentSystemPrompt],
    new MessagesPlaceholder('chat_history'),   // ← 历史消息注入这里
    ['human', '{input}'],
    new MessagesPlaceholder('agent_scratchpad'), // ← Agent 推理过程
  ]);

  const agent = await createReactAgent({
    llm: model,
    tools,
    prompt: promptWithMemory,
  });

  const executor = new AgentExecutor({
    agent,
    tools,
    memory,
    verbose: false,
    maxIterations: 5,
  });

  // 第一轮：告诉 Agent 用户的名字
  console.log('用户（第1轮）: 我叫小明，帮我查一下北京的天气');
  const result1 = await executor.invoke({ input: '我叫小明，帮我查一下北京的天气' });
  console.log('AI（第1轮）:', result1.output);

  // 第二轮：验证记忆是否生效
  console.log('\n用户（第2轮）: 我刚才说我叫什么名字？');
  const result2 = await executor.invoke({ input: '我刚才说我叫什么名字？' });
  console.log('AI（第2轮）:', result2.output);
  // 预期：AI 会说"你叫小明"，因为 BufferMemory 保存了第一轮
}

// ─────────────────────────────────────────────
// 5. 流式 Agent 输出
// ─────────────────────────────────────────────

/**
 * streamEvents 流式输出
 *
 * 可以监听 Agent 每一步的事件：
 *   on_tool_start   → 工具开始执行
 *   on_tool_end     → 工具执行完毕
 *   on_llm_stream   → AI 正在生成文字
 *   on_chain_end    → 最终结果
 *
 * 类比：直播比赛的实况解说，而不是等比赛结束再看录像
 */
async function demo3_streamingAgent() {
  console.log('\n=== 3. 流式 Agent 输出（streamEvents）===');

  const agent = await createReactAgent({
    llm: model,
    tools,
    messageModifier: agentSystemPrompt,
  });

  const executor = new AgentExecutor({
    agent,
    tools,
    verbose: false,
    maxIterations: 5,
  });

  console.log('用户: 帮我计算 123 * 456，同时查一下上海的天气\n');

  const eventStream = executor.streamEvents(
    { input: '帮我计算 123 * 456，同时查一下上海的天气' },
    { version: 'v2' }
  );

  for await (const event of eventStream) {
    if (event.event === 'on_tool_start') {
      console.log(`[工具调用] ${event.name}`, '输入:', JSON.stringify(event.data.input));
    } else if (event.event === 'on_tool_end') {
      console.log(`[工具返回] ${event.data.output}`);
    } else if (event.event === 'on_chain_end' && event.name === 'AgentExecutor') {
      console.log('\n[最终答案]', event.data.output.output);
    }
  }
}

// ─────────────────────────────────────────────
// 主函数
// ─────────────────────────────────────────────

async function main() {
  console.log('╔═══════════════════════════════════════════╗');
  console.log('║  12_langchain_agents.js  LangChain Agents  ║');
  console.log('╚═══════════════════════════════════════════╝');

  try {
    await demo1_basicAgent();
    await demo2_agentWithMemory();
    await demo3_streamingAgent();
  } catch (error) {
    console.error('演示出错:', error.message);
    if (error.message.includes('Cannot find package') || error.message.includes('Module not found')) {
      console.error('\n请先安装依赖:');
      console.error('  npm install @langchain/anthropic @langchain/core langchain zod');
    }
  }

  console.log('\n所有演示完成。');
}

main().catch(err => {
  console.error('未捕获异常:', err.message);
  process.exit(1);
});

File: code/examples/python/12_langchain_agents.py
"""
LangChain Agent 完整示例
=========================
本文件演示 LangChain Agent 的核心用法，包括：
1. @tool 装饰器定义工具
2. BaseTool 子类（复杂工具）
3. create_react_agent（内置 ReAct 策略）
4. AgentExecutor 执行器
5. ConversationBufferMemory 对话记忆
6. 流式 Agent 输出
7. LangSmith 追踪（LANGCHAIN_TRACING_V2=true）
8. 内置工具使用（Calculator、PythonREPL 等）

安装依赖：
    pip install langchain langchain-openai langchain-community langchain-experimental

环境变量（可选）：
    OPENAI_API_KEY=sk-...          # LLM 调用密钥
    LANGCHAIN_TRACING_V2=true      # 开启 LangSmith 追踪
    LANGCHAIN_API_KEY=ls__...      # LangSmith API 密钥
    LANGCHAIN_PROJECT=my-project   # LangSmith 项目名
"""

import os
from typing import Any, Optional, Type

# ── 1. @tool 装饰器 ──────────────────────────────────────────────────────────
# @tool 是最简单的工具定义方式：写一个普通函数，加上装饰器就变成 Agent 可用的工具。
# 函数的 docstring 会被 LLM 读取，用来决定"什么时候该调用这个工具"，所以要写清楚。

from langchain.tools import tool


@tool
def add_numbers(expression: str) -> str:
    """
    对两个整数求和。
    输入格式："数字1,数字2"，例如 "3,5"。
    返回两数之和。
    """
    try:
        parts = expression.split(",")
        a, b = int(parts[0].strip()), int(parts[1].strip())
        result = a + b
        return f"{a} + {b} = {result}"
    except Exception as e:
        return f"计算出错，请检查输入格式（正确示例：3,5）。错误：{e}"


@tool
def get_weather(city: str) -> str:
    """
    查询指定城市的当前天气。
    输入城市名称（中文或英文均可），返回模拟天气信息。
    """
    # 实际项目中这里会调用真实的天气 API，此处用模拟数据演示
    mock_weather = {
        "北京": "晴天，气温 28°C，湿度 40%，东南风 3 级",
        "上海": "多云，气温 32°C，湿度 75%，东风 2 级",
        "广州": "阵雨，气温 35°C，湿度 85%，南风 1 级",
    }
    weather = mock_weather.get(city, f"{city} 的天气数据暂不可用（仅支持北京/上海/广州）")
    return weather


@tool
def search_knowledge(query: str) -> str:
    """
    在本地知识库中搜索相关信息。
    输入搜索关键词，返回匹配到的知识条目。
    """
    # 实际项目中可接入向量数据库（如 Chroma、FAISS）做语义搜索
    knowledge_base = {
        "langchain": "LangChain 是一个用于构建 LLM 应用的 Python/JS 框架，提供链、Agent、记忆等核心抽象。",
        "agent": "Agent 是能够自主决策、调用工具、完成多步任务的 AI 系统，核心是 LLM + 工具 + 循环推理。",
        "react": "ReAct 是一种 Agent 策略：先 Reasoning（推理），再 Acting（行动），交替进行直到得出答案。",
        "memory": "LangChain Memory 组件用于在多轮对话中保存历史消息，让 LLM 能记住上下文。",
    }
    for keyword, content in knowledge_base.items():
        if keyword.lower() in query.lower():
            return content
    return f"未找到与 '{query}' 相关的知识条目。"


# ── 2. BaseTool 子类（复杂工具） ─────────────────────────────────────────────
# 当工具逻辑较复杂，需要初始化参数、状态管理或异步支持时，继承 BaseTool 更灵活。

from langchain.tools import BaseTool
from pydantic import BaseModel, Field


class FileReaderInput(BaseModel):
    """FileReaderTool 的输入参数 Schema，使用 Pydantic 定义。"""

    file_path: str = Field(description="要读取的文件路径（相对或绝对路径）")
    max_lines: int = Field(default=10, description="最多读取的行数，默认 10 行")


class FileReaderTool(BaseTool):
    """
    读取本地文本文件内容的工具。

    继承 BaseTool 的好处：
    - 可以在 __init__ 中注入依赖（如数据库连接、配置等）
    - 支持 args_schema 做严格的输入校验
    - 可以覆写 _arun 实现异步版本
    """

    name: str = "file_reader"
    description: str = (
        "读取本地文件内容。当需要查看某个文件的内容时使用此工具。"
        "输入文件路径和可选的最大行数。"
    )
    # 绑定输入 Schema，LangChain 会自动校验 Agent 传来的参数
    args_schema: Type[BaseModel] = FileReaderInput

    # BaseTool 要求实现 _run 方法（同步版本）
    def _run(self, file_path: str, max_lines: int = 10) -> str:
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                lines = f.readlines()[:max_lines]
            content = "".join(lines)
            return f"文件前 {max_lines} 行内容：\n{content}"
        except FileNotFoundError:
            return f"文件不存在：{file_path}"
        except Exception as e:
            return f"读取文件出错：{e}"

    # 覆写 _arun 实现异步版本（可选，不实现时异步调用会回退到同步版本）
    async def _arun(self, file_path: str, max_lines: int = 10) -> str:
        # 实际项目中可使用 aiofiles 做真正的异步文件读取
        return self._run(file_path, max_lines)


# ── 3. 内置工具 ──────────────────────────────────────────────────────────────
# LangChain 提供了一些开箱即用的工具，无需自己实现。

def get_builtin_tools() -> list:
    """
    加载常用内置工具。

    注意：
    - LLMMathChain 需要一个 LLM 实例来解析数学表达式
    - PythonREPLTool 会在本地执行 Python 代码，生产环境请谨慎使用
    - WikipediaQueryRun 需要安装 wikipedia 包：pip install wikipedia
    """
    tools = []

    # PythonREPL：让 Agent 可以执行 Python 代码（强大但有安全风险）
    try:
        from langchain_experimental.tools import PythonREPLTool
        python_tool = PythonREPLTool()
        tools.append(python_tool)
        print("[内置工具] PythonREPLTool 已加载")
    except ImportError:
        print("[内置工具] PythonREPLTool 未安装（需要 langchain-experimental）")

    return tools


# ── 4. 构建 Agent（create_react_agent + AgentExecutor） ──────────────────────
# create_react_agent 将 LLM、工具列表、提示词模板组合成一个 ReAct Agent。
# AgentExecutor 是运行 Agent 的"引擎"，负责循环调用 Agent 直到任务完成。

from langchain import hub
from langchain.agents import AgentExecutor, create_react_agent
from langchain_openai import ChatOpenAI


def build_agent(tools: list, verbose: bool = True) -> AgentExecutor:
    """
    构建一个基础的 ReAct Agent。

    参数：
        tools    - 工具列表，Agent 可以从中选择并调用
        verbose  - 是否打印 Agent 的思考过程（调试时非常有用）

    返回：
        AgentExecutor 实例
    """
    # 初始化 LLM（这里用 GPT-4o-mini，性价比高）
    llm = ChatOpenAI(
        model="gpt-4o-mini",
        temperature=0,          # 设为 0 让推理更确定性，减少随机性
        streaming=True,         # 开启流式，配合后面的流式输出使用
    )

    # 从 LangChain Hub 拉取标准 ReAct 提示词模板
    # 这个模板告诉 LLM：你有哪些工具、如何格式化思考过程和行动
    # 模板内容大致是：
    #   "你是一个 Agent，有以下工具：{tools}
    #    按 Thought/Action/Observation 格式输出..."
    prompt = hub.pull("hwchase17/react")

    # 组合成 Agent（注意：这里只是定义，还没开始执行）
    agent = create_react_agent(
        llm=llm,
        tools=tools,
        prompt=prompt,
    )

    # AgentExecutor 包裹 Agent，提供：
    # - max_iterations：防止无限循环（默认 15）
    # - handle_parsing_errors：LLM 输出格式错误时优雅处理而不崩溃
    # - verbose：打印每一步的 Thought/Action/Observation
    executor = AgentExecutor(
        agent=agent,
        tools=tools,
        verbose=verbose,
        max_iterations=10,
        handle_parsing_errors=True,
    )

    return executor


# ── 5. ConversationBufferMemory 对话记忆 ─────────────────────────────────────
# Memory 让 Agent 在多轮对话中记住之前说过的话。
# ConversationBufferMemory 是最简单的实现：把所有历史消息原样保存在内存里。

from langchain.memory import ConversationBufferMemory


def build_agent_with_memory(tools: list) -> AgentExecutor:
    """
    构建带有对话记忆的 Agent。

    与基础 Agent 的区别：
    - 每次对话结束后，问题和答案都会存入 memory
    - 下次对话时，历史记录会附加到提示词中，LLM 能记住上下文
    """
    llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)

    # ConversationBufferMemory 配置说明：
    # - memory_key：在提示词模板中引用历史记录的变量名
    # - return_messages：True 时返回 Message 对象列表（适合 ChatModel）
    memory = ConversationBufferMemory(
        memory_key="chat_history",
        return_messages=True,
    )

    # 带记忆的 Agent 需要用支持 chat_history 变量的提示词模板
    # hwchase17/react-chat 是官方提供的带记忆版本
    prompt = hub.pull("hwchase17/react-chat")

    agent = create_react_agent(llm=llm, tools=tools, prompt=prompt)

    executor = AgentExecutor(
        agent=agent,
        tools=tools,
        memory=memory,          # 传入记忆组件
        verbose=True,
        max_iterations=10,
        handle_parsing_errors=True,
    )

    return executor


# ── 6. 流式 Agent 输出 ───────────────────────────────────────────────────────
# 流式输出让用户不必等待 Agent 完全结束，而是看到逐字生成的过程，体验更好。

def run_agent_streaming(executor: AgentExecutor, query: str) -> None:
    """
    以流式方式运行 Agent，实时打印每个 token。

    LangChain 的流式 Agent 通过 stream() 方法实现，
    每次 yield 一个事件字典，包含 Agent 的中间步骤或最终答案。
    """
    print(f"\n[流式执行] 问题：{query}")
    print("-" * 50)

    # stream() 返回一个生成器，每次 yield 一个事件
    for chunk in executor.stream({"input": query}):
        # chunk 可能包含以下键：
        # - "actions"：Agent 决定调用某个工具（中间步骤）
        # - "steps"：工具执行完毕，返回观察结果
        # - "output"：最终答案

        if "actions" in chunk:
            for action in chunk["actions"]:
                print(f"\n[思考] 决定调用工具: {action.tool}")
                print(f"[参数] {action.tool_input}")

        elif "steps" in chunk:
            for step in chunk["steps"]:
                print(f"\n[观察] 工具返回: {step.observation}")

        elif "output" in chunk:
            print(f"\n[最终答案] {chunk['output']}")

    print("-" * 50)


# ── 7. LangSmith 追踪配置 ────────────────────────────────────────────────────
# LangSmith 是 LangChain 官方的可观测性平台，可以记录每次 LLM 调用的输入输出、
# 工具调用链路、耗时、费用等，方便调试和性能分析。

def setup_langsmith_tracing() -> None:
    """
    配置 LangSmith 追踪。

    只需设置以下环境变量，LangChain 会自动将所有调用上报到 LangSmith：
        LANGCHAIN_TRACING_V2=true        - 开启追踪
        LANGCHAIN_API_KEY=ls__...        - 你的 LangSmith API 密钥
        LANGCHAIN_PROJECT=my-project     - 追踪数据归属的项目名

    查看追踪数据：https://smith.langchain.com
    """
    # 检查是否已通过环境变量配置
    if os.getenv("LANGCHAIN_TRACING_V2") == "true":
        project = os.getenv("LANGCHAIN_PROJECT", "default")
        print(f"[LangSmith] 追踪已开启，项目：{project}")
        print("[LangSmith] 访问 https://smith.langchain.com 查看追踪数据")
    else:
        print("[LangSmith] 追踪未开启。如需开启，请设置：")
        print("  export LANGCHAIN_TRACING_V2=true")
        print("  export LANGCHAIN_API_KEY=ls__your_key")
        print("  export LANGCHAIN_PROJECT=my-project")

    # 也可以在代码中直接设置（不推荐，密钥不应硬编码）：
    # os.environ["LANGCHAIN_TRACING_V2"] = "true"
    # os.environ["LANGCHAIN_API_KEY"] = "ls__your_key"


# ── 8. 主函数：整合所有功能演示 ──────────────────────────────────────────────

def demo_basic_tools() -> None:
    """演示用 @tool 装饰器定义的工具（不依赖 LLM，可独立测试）。"""
    print("\n=== 演示 @tool 装饰器定义的工具 ===")

    # 直接调用工具（绕过 Agent，用于单元测试）
    result = add_numbers.invoke("3,7")
    print(f"add_numbers('3,7') => {result}")

    result = get_weather.invoke("北京")
    print(f"get_weather('北京') => {result}")

    result = search_knowledge.invoke("什么是 ReAct")
    print(f"search_knowledge('什么是 ReAct') => {result}")

    # 查看工具元信息（LLM 会读取这些信息决定何时调用工具）
    print(f"\n工具名称: {add_numbers.name}")
    print(f"工具描述: {add_numbers.description}")


def demo_base_tool() -> None:
    """演示 BaseTool 子类的使用（不依赖 LLM，可独立测试）。"""
    print("\n=== 演示 BaseTool 子类 ===")

    reader = FileReaderTool()

    # 读取一个不存在的文件，测试错误处理
    result = reader.invoke({"file_path": "/tmp/not_exist.txt", "max_lines": 5})
    print(f"读取不存在的文件: {result}")

    # 创建临时文件并读取
    tmp_path = "/tmp/langchain_demo.txt"
    with open(tmp_path, "w") as f:
        f.write("第1行：LangChain 是构建 LLM 应用的框架\n")
        f.write("第2行：Agent 可以自主调用工具完成任务\n")
        f.write("第3行：Memory 让对话有上下文记忆\n")

    result = reader.invoke({"file_path": tmp_path, "max_lines": 2})
    print(f"读取临时文件（前2行）:\n{result}")


def demo_full_agent() -> None:
    """
    完整 Agent 演示（需要 OPENAI_API_KEY 环境变量）。

    如果没有 API Key，这个函数会打印提示信息后跳过。
    """
    print("\n=== 完整 Agent 演示 ===")

    if not os.getenv("OPENAI_API_KEY"):
        print("[跳过] 未设置 OPENAI_API_KEY，跳过需要 LLM 的演示。")
        print("  设置方法：export OPENAI_API_KEY=sk-...")
        return

    # 检查 LangSmith 配置
    setup_langsmith_tracing()

    # 组合工具列表
    tools = [
        add_numbers,        # @tool 装饰器定义的工具
        get_weather,        # @tool 装饰器定义的工具
        search_knowledge,   # @tool 装饰器定义的工具
        FileReaderTool(),   # BaseTool 子类定义的工具
    ]

    # 尝试加载内置工具
    tools.extend(get_builtin_tools())

    print(f"\n已加载 {len(tools)} 个工具：{[t.name for t in tools]}")

    # --- 基础 Agent（单轮问答）---
    print("\n--- 基础 Agent（单轮问答）---")
    executor = build_agent(tools, verbose=True)

    # Agent 会自动选择合适的工具来回答问题
    response = executor.invoke({"input": "北京今天天气怎么样？"})
    print(f"\n最终答案：{response['output']}")

    # --- 流式 Agent ---
    print("\n--- 流式 Agent 输出 ---")
    run_agent_streaming(executor, "请帮我计算 42 加上 58 等于多少？")

    # --- 带记忆的 Agent（多轮对话）---
    print("\n--- 带记忆的 Agent（多轮对话）---")
    memory_executor = build_agent_with_memory(tools)

    # 第一轮：问天气
    r1 = memory_executor.invoke({"input": "上海今天天气如何？"})
    print(f"第一轮答案：{r1['output']}")

    # 第二轮：引用上轮对话（Agent 应该记得"上海天气"的上下文）
    r2 = memory_executor.invoke({"input": "那广州呢？和前面那个城市比怎么样？"})
    print(f"第二轮答案：{r2['output']}")


# ── 入口 ─────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    print("=" * 60)
    print("LangChain Agent 完整示例")
    print("=" * 60)

    # 步骤1：演示工具定义（无需 API Key）
    demo_basic_tools()

    # 步骤2：演示 BaseTool 子类（无需 API Key）
    demo_base_tool()

    # 步骤3：完整 Agent 演示（需要 OPENAI_API_KEY）
    demo_full_agent()

    print("\n演示完成。")
    print("提示：设置 OPENAI_API_KEY 后可运行完整 Agent 演示。")
